sys/fido2/ctap/ctap_mem.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2021 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup fido2_ctap_mem
+ * @{
+ * @file
+ *
+ * @author      Nils Ollrogge <nils.ollrogge@fu-berlin.de>
+ * @}
+ */
+
+#include <string.h>
+
+#include "mtd.h"
+#include "mtd_flashpage.h"
+
+#include "fido2/ctap/ctap_mem.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   MTD device descriptor initialized with flash-page driver
+ */
+static mtd_dev_t _mtd_dev = MTD_FLASHPAGE_INIT_VAL(1);
+
+/**
+ * @brief   Max amount of resident keys that can be stored
+ */
+static uint16_t _max_rk_amnt;
+
+/**
+ * @brief   Check if flash region is erased
+ */
+static bool _flash_is_erased(int page, int offset, size_t len);
+
+int fido2_ctap_mem_init(void)
+{
+    int ret;
+
+    ret = mtd_init(&_mtd_dev);
+
+    if (ret < 0) {
+        return ret;
+    }
+
+    for (unsigned i = CTAP_FLASH_RK_START_PAGE; i < FLASHPAGE_NUMOF; i++) {
+        _max_rk_amnt += flashpage_size(i) / CTAP_FLASH_RK_SZ;
+    }
+
+    return CTAP2_OK;
+}
+
+int fido2_ctap_mem_read(void *buf, uint32_t page, uint32_t offset, uint32_t len)
+{
+    assert(buf);
+
+    int ret;
+
+    ret = mtd_read_page(&_mtd_dev, buf, page, offset, len);
+
+    if (ret < 0) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    return CTAP2_OK;
+}
+
+int fido2_ctap_mem_write(const void *buf, uint32_t page, uint32_t offset, uint32_t len)
+{
+    assert(buf);
+
+    int ret;
+
+    if (!_flash_is_erased(page, offset, len)) {
+        ret = mtd_write_page(&_mtd_dev, buf, page, offset, len);
+
+        if (ret < 0) {
+            return CTAP1_ERR_OTHER;
+        }
+    }
+    else {
+        ret = mtd_write_page_raw(&_mtd_dev, buf, page, offset, len);
+
+        if (ret < 0) {
+            return CTAP1_ERR_OTHER;
+        }
+    }
+
+    return CTAP2_OK;
+}
+
+static bool _flash_is_erased(int page, int offset, size_t len)
+{
+    uint8_t *addr = ((uint8_t *)flashpage_addr(page) + offset);
```

Comment:
I just noticed that `mtd_flashpage` uses `FLAGPAGE_SIZE` for page_size instead of the `flagpage_size()`function. I am pretty sure this can break my code if `FLAGPAGE_SIZE` != `flagpage_size(i)` ...